File: arena-junit/src/test/java/arena/junit/ChildrenLifecycleComponentTest.java
package arena.junit;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import arena.junit.exec.ExecutableComponent;
import arena.junit.exec.ExecutableComponentBuilder;
import arena.junit.match.Match;
import arena.junit.match.MatchBuilder;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.Test;

final class ChildrenLifecycleComponentTest {

  private static final String MATCH_NAME = "children-lifecycle-probe";

  @Test
  void open_withChildComponent_startsBothParentAndChild() throws Exception {
    Path markerFile = Files.createTempFile("arena-junit-children", ".txt");
    try {
      ExecutableComponent child =
          new ExecutableComponentBuilder("child")
              .withExecutablePath("/bin/sh")
              .withRuntimeArg("flag", "-c")
              .withRuntimeArg("script", "echo child >> " + markerFile)
              .build();

      ExecutableComponent parent =
          new ExecutableComponentBuilder("parent")
              .withExecutablePath("/bin/sh")
              .withRuntimeArg("flag", "-c")
              .withRuntimeArg("script", "echo parent >> " + markerFile)
              .addChildComponent(child)
              .build();

      Match match = new MatchBuilder(MATCH_NAME).addComponent(parent).build();
      ClosedArena closedArena = new ClosedArena(MATCH_NAME, List.of(match));
      OpenArena arena = closedArena.open();
      try {
        assertNotNull(arena);
        List<String> lines = waitForMarkerLines(markerFile, 5000);
        assertTrue(lines.contains("child"), lines::toString);
        assertTrue(lines.contains("parent"), lines::toString);
      } finally {
        arena.close();
      }
    } finally {
      Files.deleteIfExists(markerFile);
    }
  }

  private static List<String> waitForMarkerLines(Path markerFile, long timeoutMs)
      throws IOException, InterruptedException {
    long deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs);
    while (true) {
      List<String> lines = Files.readAllLines(markerFile);
      if (lines.contains("child") && lines.contains("parent")) {
        return lines;
      }
      if (System.nanoTime() >= deadline) {
        return lines;
      }
      TimeUnit.MILLISECONDS.sleep(20);
    }
  }
}

File: arena-pytest/tests/component/children/children_lifecycle_component_test.py
from __future__ import annotations

import asyncio
import os
import tempfile
import time
from typing import List

from arena_pytest import ClosedArena, ExecutableComponentBuilder, MatchBuilder

MATCH_NAME = "children-lifecycle-probe"


async def _open_and_close_with_children(marker_file: str) -> None:
    child = (
        ExecutableComponentBuilder("child")
        .with_executable_path("/bin/sh")
        .with_runtime_arg("flag", "-c")
        .with_runtime_arg("script", f"echo child >> {marker_file}")
        .build()
    )
    parent = (
        ExecutableComponentBuilder("parent")
        .with_executable_path("/bin/sh")
        .with_runtime_arg("flag", "-c")
        .with_runtime_arg("script", f"echo parent >> {marker_file}")
        .with_child_components([child])
        .build()
    )
    a_match = MatchBuilder(MATCH_NAME).add_component(parent).build()
    closed = ClosedArena(MATCH_NAME, [a_match])
    arena = await closed.open()
    try:
        assert arena is not None
    finally:
        await arena.close()


def _wait_for_marker_lines(marker_file: str, timeout_s: float = 5.0) -> List[str]:
    deadline = time.monotonic() + timeout_s
    while time.monotonic() < deadline:
        if os.path.exists(marker_file):
            with open(marker_file) as f:
                lines = [line.strip() for line in f.readlines()]
            if "child" in lines and "parent" in lines:
                return lines
        time.sleep(0.02)
    return []


def test_open_arena_with_child_component_starts_both_parent_and_child():
    with tempfile.TemporaryDirectory() as tmp_dir:
        marker_file = os.path.join(tmp_dir, "marker.txt")
        asyncio.run(_open_and_close_with_children(marker_file))
        lines = _wait_for_marker_lines(marker_file)
        assert "child" in lines
        assert "parent" in lines
